### ISSUE TITLE
Example of callback functions?

### ISSUE THREAD

**Issue by zeno17** [UKPLab/sentence-transformers#458]
Hi,

I was wondering if and where i could find callback functions to be used during training, e.g. i know Tensorflow has EarlyStopping?

**Comment by nreimers** [comment]
Have a look here:
https://github.com/UKPLab/sentence-transformers/pull/327

**Comment by zeno17** [comment]
@PhilipMay 

1. I think the docs were updated to https://optuna.readthedocs.io/en/stable/tutorial/007_pruning.html (so your link is dead).
2. Do you by any chance have example of code that you can share on how i could use Optuna callbacks to optimize a sentence transformer?

**Comment by PhilipMay** [comment]
I just do this very simple thing:

```python
def callback(value, a, b):
    print('callback:', value, a, b)
    if math.isnan(value):
        raise optuna.exceptions.TrialPruned()

    model.fit(train_objectives=[(train_dataloader, train_loss)],
              evaluator=dev_evaluator,
              epochs=num_epochs,
              scheduler=trial.suggest_categorical('scheduler', ['WarmupLinear', 'warmupcosine', 'warmupcosinewithhardrestarts']),
              #optimizer_class=optimizer_class,
              evaluation_steps=evaluation_steps,
              warmup_steps=warmup_steps,
              output_path=model_save_path,
              optimizer_params={'lr': lr, 'eps': eps, 'correct_bias': False},
              weight_decay=weight_decay,
              callback=callback,
              )
```

But since a full training needs many houry I can use this callback to save a lot of time.

**Comment by zeno17** [comment]
Could you post your objective function aswell?

**Comment by PhilipMay** [comment]
@zeno17 here is just the full code - feedback welcome. :-)

**Comment by milmin** [comment]
do you have any example of a valid callback for tracking train and validation loss? (sorry, I'm quite new to pytorch...)

**Comment by nreimers** [comment]
Hi @milmin 
You can use the evaluator parameter and pass multiple evaluators by wrapping them in a evaluation.SequentialEvaluator(). There you can compute your various validation scores you want to track.

**Comment by milmin** [comment]
Thanks for your reply, is there already such an example in the doc? It would be definitely very useful for newcomers by the way :-)

**Comment by nreimers** [comment]
The usage is rather simple:
```
model.fit(evaluator=evaluation.SequentialEvaluator([evaluator1, evaluator2, evaluator3]), ...)
```

with evaluator1/2/3 your different evaluators.

Note, by default, the value returned by the last evaluator will be used if a new model checkpoint will be saved.

**Comment by milmin** [comment]
What about if I want to track train metrics as well?

**Comment by milmin** [comment]
I tried something like `evaluation.SequentialEvaluator([train_evaluator, dev_evaluator])` to track the train and validation scores at least at the evaluation steps, however the evaluator returns only one value: as described in the doc 

> All scores are passed to ‘main_score_function’, which derives one final score value

how is this final score value computed? Can you give some more details? I think it would be very useful to have one value for every evaluator at each evaluation step.

**Comment by tempdeltavalue** [comment]
Can I get training model inside of custom callback function ? 
Like in keras: 
https://stackoverflow.com/questions/65302030/access-loss-and-model-in-a-custom-callback
